### ISSUE TITLE
Sometimes Deadlock from TF Dataset using Generator

### ISSUE THREAD

**Issue by ghostplant** [tensorflow/tensorflow#22632]
Seems to be a normal generator code, but 50% possibility to hang on exit / or print deconstruction exceptions, and it is also possible to exit successfully.
The Tensorflow version I use v1.10

```sh
import tensorflow as tf
import numpy as np

data_dir, batch_size, depth, height, width = '/tmp/dataset/flowers_jpegs', 64, 3, 224, 224

def generator():
  from keras.preprocessing.image import ImageDataGenerator
  from keras.utils import OrderedEnqueuer
  gen = ImageDataGenerator(data_format='channels_first', rescale=1./255, fill_mode='nearest').flow_from_directory(
                           data_dir + '/train', target_size=(height, width), batch_size=batch_size)
  enqueuer = OrderedEnqueuer(gen, use_multiprocessing=False)
  enqueuer.start(workers=16)
  n_classes = gen.num_classes

  while True:
    batch_xs, batch_ys = next(enqueuer.get())
    yield batch_xs, batch_ys

ds = tf.data.Dataset.from_generator(generator, (tf.float32, tf.float32))
ds = ds.prefetch(buffer_size=batch_size)
ds_iter = ds.make_one_shot_iterator()


with tf.Session() as sess:
  images, labels = ds_iter.get_next()
  images = tf.reshape(images, (-1, 3, height, width))
  labels = tf.reshape(labels, (-1, 2))
  print(sess.run(images).reshape([-1])[0:8])

  images, labels = ds_iter.get_next()
  images = tf.reshape(images, (-1, 3, height, width))
  labels = tf.reshape(labels, (-1, 2))
  print(sess.run(images).reshape([-1])[0:8])
```

**Environment:**
**Have I written custom code:** The code above the the complete single python3 file, and no other libraries include customized changes.
**OS Platform and Distribution: Ubuntu 16.04 x86-64 LTS;
**TensorFlow installed from:** clone `tensorflow:v1.10.1` from github and build for CUDA 1.10 and CUDNN 7.3;
**Bazel version:** 0.15.0
**CUDA/cuDNN version:** CUDA = 10.0; CUDNN = 7.3;
**GPU model and memory:** Tesla V100 (each GPU 16 GB memory)
**Exact command to reproduce:** `python3 above-code.py` (note that "/tmp/dataset/flowers_jpegs" is a standard directory including original jpeg images inside serveral subfolders (as different classes) which satisfies Keras ImageData directory input format.)
**Mobile device:** N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by ghostplant** [comment]
Thanks you, I have already updated these environment information.

**Comment by ymodak** [comment]
Can you please provide a link to download data_dir as well?
Thanks

**Comment by ghostplant** [comment]
You can create a simple data_dir by the following steps:

```sh
1) mkdir -p /tmp/dataset/flowers_jpegs/train/{rose,daisy}
2) then copy at least 1 or 2 image file (*.jpeg) to each sub directory
3) done.
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ghostplant** [comment]
Seems like Tensorflow is easy to have an unexpected exception when the tensorflow session exits.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ghostplant** [comment]
I close this issue because tf.data.Dataset.from_generator is not a good choice for HP training.
